Add Playwright E2E tests boilerplates

jakule opened the pull request:
Why?
---

Teleport lacks E2E, which would allow catching more WebUI errors and potential regression in WebUI <-> backend communication.

What
---

This commit introduces end-to-end tests with Docker Compose and Playwrite to improve code quality and provide a more robust testing environment. This involves adding a GitHub workflow for manually triggering the test suite, Makefile commands for running, the tests, and configurations. This addition will enable easier testing and provide a platform for future test development.

The goal
---

This PR is not meant to be a new Teleport-wide solution for E2E testing but rather an MVP for testing basic functionality. If the approach works on a small scale, we can extend the test coverage to other services like DB and K8S. We want to focus on SSH and Assist and collect feedback for now. CI will run only on the manual trigger.

I decided to use Playwright and Docker Compose for the MVP. 
Playwright allows writing tests easier with auto-code generation.
Docker Compose should help set up an isolated and repeatable environment for local development and CI.

jakule committed:
Add end-to-end tests with Playwright

This commit introduces end-to-end tests with Docker Compose to improve code quality and provide a more robust testing environment. This involves adding a GitHub workflow for manually triggering the test suite, Makefile commands for running, the tests, and configurations. This addition will enable easier testing and provide a platform for future test development.

jakule committed:
Cleanup

jakule committed:
Only allow manual CI trigger

jakule committed:
Ignore e2e tests in Jest configuration

Added 'testPathIgnorePatterns' field to the Jest configuration in order to ignore end-to-end tests when running unit tests.

jakule commented:
Friendly ping @ryanclark @zmb3 @r0mant

jakule commented on e2e/config/create-user.sh:
```diff
@@ -0,0 +1,40 @@
+tctl -d users add bob --roles=access,editor --logins=root
+# Set the password - 'secret'.
+# TODO(jakule): I was not able to figure out how to set the password using WebUI in automated way, so I'm doing it directly in the database.
+sqlite3 /var/lib/teleport/backend/sqlite.db "UPDATE kv SET value='\$2a\$10\$w0K2pwK/cF8BG0kKZ7X1qe1uU7w3mwZ2S46PO6SlaiVjiTkbNGQp6' where key = '/web/users/bob/pwd';"
```
I'll change that to use `tctl create` to create users.

jakule replied:
Not needed after applying this suggestion https://github.com/gravitational/teleport/pull/29286#discussion_r1273864120

jakule commented on e2e/config/create-user.sh:
Sure, it sounds like a much better solution. I was not aware that we have something like that :sweat_smile:

jakule replied:
Changed

jakule commented on .github/workflows/e2e-tests.yaml:
```diff
@@ -0,0 +1,62 @@
+name: E2E Tests
+run-name: E2E Tests
+on:
+  # Only allow manual triggers
```
Long story. Currently, this job is not doing anything useful, and building the binaries takes ages. A lot of $$ and no value. From my experience, E2E tests get slow very quickly.
If you make changes to this file itself you can trigger the build manually from a branch. I'd leave as it is right now and revisit this topic when we have more tests.

jakule commented on e2e/config/teleport.yaml:
```diff
@@ -0,0 +1,45 @@
+
+ssh_service:
+  enabled: "yes"
+  public_addr: teleport:3080
```
Works for me, but you're right. The port should be different.

jakule replied:
Changed

sshahcodes commented on e2e/tests/assist.spec.ts:
```diff
@@ -0,0 +1,30 @@
+  await page.getByPlaceholder('Username').fill('bob');
+  await page.getByPlaceholder('Password').fill('secret');
+
+  await page.getByRole('button', { name: 'Sign In' }).click();
```
You may have already considered this but just calling it out, Playwright supports shared authentication state: https://playwright.dev/docs/auth. Unless we are testing for the sign in functionality itself, we can set up shared auth and have rest of the tests written in a logged in context.

jakule replied:
I'm aware of that, but the current implementation is fine for the one test. After this PR is merged we can optimize it.

flyinghermit reviewed (commented):
I like it we are setting up e2e tests but also a bit wary on additional time it would add to GA runners (I understand this requires manual trigger for now). I have used puppeteer before and timeouts used to be a problem when running in CI. Maybe only run e2e before release?

jakule commented:
> I like it we are setting up e2e tests but also a bit wary on additional time it would add to GA runners (I understand this requires manual trigger for now). I have used puppeteer before and timeouts used to be a problem when running in CI. Maybe only run e2e before release?

As I mentioned above. E2E tests are expensive and slow to run. I don't think that running them before each release is a good idea, as if you realize that something is broken on the release day, the release probably will be delayed. Currently, I'm leaning towards running them nightly, but IMO we should start with something. It should be easy to run them locally and on GHA when needed. We can see what the situation looks like after we add more tests and then decide.
The bottom line, let's get something running first and then optimize. Not the other way around.

jakule committed:
Address code review comments

jakule commented on .github/workflows/e2e-tests.yaml:
```diff
@@ -0,0 +1,62 @@
+          make -C e2e build
+      - name: Run Tests
+        run: |
+          make -C e2e test
```
Changed

jakule commented:
Friendly ping @rosstimothy @r0mant @zmb3

kimlisa commented on e2e/playwright.config.ts:
```diff
@@ -0,0 +1,83 @@
+      use: { ...devices['Desktop Chrome'] },
+    },
+
+    // {
+    //   name: 'firefox',
+    //   use: { ...devices['Desktop Firefox'] },
+    // },
```
why disable the other major browsers?

jakule replied:
Because I'm not sure how compatible we're with other major browsers, and I also don't want the tests to take forever to run. I think we can revisit this when we have something working.

rosstimothy reviewed (commented):
This does not work locally for me.

```bash
$ cd e2e

$ make build

docker compose build
[+] Building 0.2s (11/11) FINISHED
 => [teleport internal] load build definition from Dockerfile                                                                                                                             0.0s
 => => transferring dockerfile: 801B                                                                                                                                                      0.0s
 => [teleport internal] load .dockerignore                                                                                                                                                0.0s
 => => transferring context: 2B                                                                                                                                                           0.0s
 => [teleport internal] load metadata for docker.io/library/ubuntu:22.04                                                                                                                  0.2s
 => [teleport 1/6] FROM docker.io/library/ubuntu:22.04@sha256:0bced47fffa3361afa981854fcabcd4577cd43cebbb808cea2b1f33a3dd7f508                                                            0.0s
 => [teleport internal] load build context                                                                                                                                                0.0s
 => => transferring context: 95B                                                                                                                                                          0.0s
 => CACHED [teleport 2/6] RUN apt-get update &&     apt-get install -y     curl     ca-certificates     && rm -rf /var/lib/apt/lists/*                                                    0.0s
 => CACHED [teleport 3/6] RUN curl -JLO "https://dl.filippo.io/mkcert/latest?for=linux/amd64" &&     chmod +x mkcert-v*-linux-amd64 &&     cp mkcert-v*-linux-amd64 /usr/local/bin/mkcer  0.0s
 => CACHED [teleport 4/6] RUN mkdir -p /etc/teleport/certs &&     mkcert -install &&     mkcert -cert-file /etc/teleport/certs/tls.crt -key-file /etc/teleport/certs/tls.key localhost    0.0s
 => CACHED [teleport 5/6] COPY ./build/teleport /usr/local/bin/teleport                                                                                                                   0.0s
 => CACHED [teleport 6/6] COPY ./build/tctl /usr/local/bin/tctl                                                                                                                           0.0s
 => [teleport] exporting to image                                                                                                                                                         0.0s
 => => exporting layers                                                                                                                                                                   0.0s
 => => writing image sha256:d4bd5fa73f4bd9411d9c03ea6751d175ad3e166b4912a7cf98551c0ee2ffe63c                                                                                              0.0s
 => => naming to docker.io/library/teleport-e2e:latest                                                                                                                                    0.0s
docker compose create
[+] Creating 2/0
 ✔ Container e2e-teleport-1  Created                                                                                                                                                      0.0s
 ✔ Container e2e-e2e-1       Created 

$ make test
docker compose up --abort-on-container-exit --exit-code-from e2e
[+] Running 2/0
 ✔ Container e2e-teleport-1  Created                                                                                                                                                      0.0s
 ✔ Container e2e-e2e-1       Created                                                                                                                                                      0.0s
Attaching to e2e-e2e-1, e2e-teleport-1
e2e-e2e-1       | Updating certificates in /etc/ssl/certs...
e2e-teleport-1  | Cannot run macOS (Mach-O) executable in Docker: Exec format error
e2e-teleport-1 exited with code 126
Aborting on container exit...
[+] Stopping 2/2
 ✔ Container e2e-e2e-1       Stopped                                                                                                                                                     10.2s
 ✔ Container e2e-teleport-1  Stopped

jakule commented on e2e/config/run-tests.sh:
```diff
@@ -0,0 +1,31 @@
+
+yarn install
+
+# Wait for the user to be created
+sleep 5
```
Yes, in some cases the test runner can start too early and fail as Teleport is not up. We could check if the Teleport responds on a port, but the fact that Teleport accepts connection doesn't mean that it's fully up.
I'm open to ideas on how to improve it, but for now I think that sleep is fine.

jakule committed:
Update e2e test environment for multi-architecture support

Modified the end-to-end test setup scripts and docker files to support both Linux and MacOS architectures. The build process now detects the system architecture and downloads the appropriate version of `mkcert`. Also, there is a control flow to build binaries only if they don't exist and the build files are now mounted from the build directory instead of being copied. These changes aim to make the e2e tests more robust and adaptable to different development environments.

jakule commented on e2e/config/teleport.yaml:
```diff
@@ -0,0 +1,44 @@
+version: v2
+teleport:
+  nodename: teleport-e2e
+  auth_token: foo
+  data_dir: /var/lib/teleport
+  log:
+    output: stderr
+    severity: TRACE
+    format:
+      output: text
+
+  auth_servers:
+    - teleport:3025
```
No reason. My dev config is still using v2, so I just copied it.

jakule committed:
Update Makefile and teleport.yaml for testing improvements

Continued refinement of testing process by updating the Makefile and teleport.yaml. Changes to the Makefile include additional phony targets, modification of build-binaries, and a new 'all' target which runs key steps in sequence. The teleport.yaml file was updated to version v3.

jakule committed:
Fix makefile on MacOS

jakule commented:
> This does not work locally for me.
> 
> ```shell
> $ cd e2e
> 
> $ make build
> 
> docker compose build
> [+] Building 0.2s (11/11) FINISHED
>  => [teleport internal] load build definition from Dockerfile                                                                                                                             0.0s
>  => => transferring dockerfile: 801B                                                                                                                                                      0.0s
>  => [teleport internal] load .dockerignore                                                                                                                                                0.0s
>  => => transferring context: 2B                                                                                                                                                           0.0s
>  => [teleport internal] load metadata for docker.io/library/ubuntu:22.04                                                                                                                  0.2s
>  => [teleport 1/6] FROM docker.io/library/ubuntu:22.04@sha256:0bced47fffa3361afa981854fcabcd4577cd43cebbb808cea2b1f33a3dd7f508                                                            0.0s
>  => [teleport internal] load build context                                                                                                                                                0.0s
>  => => transferring context: 95B                                                                                                                                                          0.0s
>  => CACHED [teleport 2/6] RUN apt-get update &&     apt-get install -y     curl     ca-certificates     && rm -rf /var/lib/apt/lists/*                                                    0.0s
>  => CACHED [teleport 3/6] RUN curl -JLO "https://dl.filippo.io/mkcert/latest?for=linux/amd64" &&     chmod +x mkcert-v*-linux-amd64 &&     cp mkcert-v*-linux-amd64 /usr/local/bin/mkcer  0.0s
>  => CACHED [teleport 4/6] RUN mkdir -p /etc/teleport/certs &&     mkcert -install &&     mkcert -cert-file /etc/teleport/certs/tls.crt -key-file /etc/teleport/certs/tls.key localhost    0.0s
>  => CACHED [teleport 5/6] COPY ./build/teleport /usr/local/bin/teleport                                                                                                                   0.0s
>  => CACHED [teleport 6/6] COPY ./build/tctl /usr/local/bin/tctl                                                                                                                           0.0s
>  => [teleport] exporting to image                                                                                                                                                         0.0s
>  => => exporting layers                                                                                                                                                                   0.0s
>  => => writing image sha256:d4bd5fa73f4bd9411d9c03ea6751d175ad3e166b4912a7cf98551c0ee2ffe63c                                                                                              0.0s
>  => => naming to docker.io/library/teleport-e2e:latest                                                                                                                                    0.0s
> docker compose create
> [+] Creating 2/0
>  ✔ Container e2e-teleport-1  Created                                                                                                                                                      0.0s
>  ✔ Container e2e-e2e-1       Created 
> 
> $ make test
> docker compose up --abort-on-container-exit --exit-code-from e2e
> [+] Running 2/0
>  ✔ Container e2e-teleport-1  Created                                                                                                                                                      0.0s
>  ✔ Container e2e-e2e-1       Created                                                                                                                                                      0.0s
> Attaching to e2e-e2e-1, e2e-teleport-1
> e2e-e2e-1       | Updating certificates in /etc/ssl/certs...
> e2e-teleport-1  | Cannot run macOS (Mach-O) executable in Docker: Exec format error
> e2e-teleport-1 exited with code 126
> Aborting on container exit...
> [+] Stopping 2/2
>  ✔ Container e2e-e2e-1       Stopped                                                                                                                                                     10.2s
>  ✔ Container e2e-teleport-1  Stopped   
> ```

Fixed. It should work on MacOS now too.

rosstimothy reviewed (commented):
This still fails for me if I already have a `build/teleport` binary that was generated from my local machine. I've only been able to get this to work by running `make clean` and then `make build`.

```
$ make build
docker compose build --build-arg BUILDARCH=amd64
[+] Building 0.5s (8/8) FINISHED
 => [teleport internal] load build definition from Dockerfile                                                                                                                             0.1s
 => => transferring dockerfile: 751B                                                                                                                                                      0.0s
 => [teleport internal] load .dockerignore                                                                                                                                                0.0s
 => => transferring context: 2B                                                                                                                                                           0.0s
 => [teleport internal] load metadata for docker.io/library/ubuntu:22.04                                                                                                                  0.4s
 => [teleport 1/4] FROM docker.io/library/ubuntu:22.04@sha256:0bced47fffa3361afa981854fcabcd4577cd43cebbb808cea2b1f33a3dd7f508                                                            0.0s
 => CACHED [teleport 2/4] RUN apt-get update &&     apt-get install -y     curl     ca-certificates     && rm -rf /var/lib/apt/lists/*                                                    0.0s
 => CACHED [teleport 3/4] RUN curl -JLO "https://dl.filippo.io/mkcert/latest?for=linux/amd64" &&     chmod +x mkcert-v*-linux-amd64 &&     cp mkcert-v*-linux-amd64 /usr/local/bin/mkcer  0.0s
 => CACHED [teleport 4/4] RUN mkdir -p /etc/teleport/certs &&     mkcert -install &&     mkcert -cert-file /etc/teleport/certs/tls.crt -key-file /etc/teleport/certs/tls.key localhost    0.0s
 => [teleport] exporting to image                                                                                                                                                         0.0s
 => => exporting layers                                                                                                                                                                   0.0s
 => => writing image sha256:00305bf48482e1a98724587a03a8c9d0c5259b580edb37970fd9fb1435c4a7e4                                                                                              0.0s
 => => naming to docker.io/library/teleport-e2e:latest                                                                                                                                    0.0s
docker compose create
[+] Creating 2/0
 ✔ Container e2e-e2e-1       Created                                                                                                                                                      0.0s
 ✔ Container e2e-teleport-1  Created     

$ make test
# Build binaries if they don't exist
test -f "../build/teleport" || make build-binaries
docker compose up --abort-on-container-exit --exit-code-from e2e
[+] Running 2/0
 ✔ Container e2e-teleport-1  Created                                                                                                                                                      0.0s
 ✔ Container e2e-e2e-1       Created                                                                                                                                                      0.0s
Attaching to e2e-e2e-1, e2e-teleport-1
e2e-teleport-1  | Cannot run macOS (Mach-O) executable in Docker: Exec format error
e2e-e2e-1       | Updating certificates in /etc/ssl/certs...
e2e-teleport-1 exited with code 126
Aborting on container exit...
[+] Stopping 1/2
 ✔ Container e2e-teleport-1  Stopped                                                                                                                                                      0.0s
 ⠙ Container e2e-e2e-1       Stopping   
```

jakule committed:
Add Readme

jakule commented:
Friendly ping @zmb3 @r0mant 
I need admin approval on this.

GavinFrazar commented:
it appears this installs the mkcert CA inside the docker image. We should install the CA as part of a workflow instead - otherwise the CA private key is distributed inside our buildbox image.
